README.md
# Blinkenlights Colour

This is my repository of the basic software for the Blinkenlights project ( http://blinkenlights.net/ ). The files are taken directly from the download links on that site ( http://blinkenlights.net/project/developer-tools ).

*The same software also exists in a GitHub repository ( https://github.com/blinkenlights/blinkenlights ) but that version lacks the files ( `configure` etc. ) needed to build the software. There are some build instructions but they seem to refer to a CVS checkout which I have no further details of. As far as I can see the code files are identical to those on the Blinkenlights website.*

I have made some additions to the basic code:

- Enable colour packets to be processed correctly.

Build instructions are contained in the README files in each section.

# Simple Test Setup

Once the software has been built and installed then the following is a very simple test to check for correct operation:

- `blinkensim -l` gives a list of themes available
- `blinkensim hdl-640x480` starts the simulator in a new window ( *Haus des Lehrers (Berlin) (18x8) on 640x480* )
- Some test movies can be found in `blib-1.1.7/test/movies`
- `bsend camel-18x8.blm` - shows an animation in the simulator window (  *add the "-l" option to loop the movie* )

Most commands have basic help by using the `-?` option


blinkenthemes-0.10/NEWS
0.10
----
- HDL theme in 320x240
- HDL2 themes in various sizes
- reset HDL themes to maxval 1
- cleanup


blib-1.1.7/blib/breceiver.c
/* blib - Library of useful things to hack the Blinkenlights
 *
 * Copyright (c) 2001-2003  The Blinkenlights Crew
 *                          Sven Neumann <sven@gimp.org>
 *                          Michael Natterer <mitch@gimp.org>
 *                          Daniel Mack <daniel@yoobay.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "config.h"

#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include <glib-object.h>

#ifdef G_OS_WIN32
#include <winsock2.h>
#ifndef socklen_t
#define socklen_t unsigned int
#endif
#else
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#endif

#include "btypes.h"
#include "bpacket.h"
#include "breceiver.h"
#include "bsocket.h"


enum
{
  PROP_0,
  PROP_CALLBACK,
  PROP_CALLBACK_DATA
};


static void     b_receiver_class_init   (BReceiverClass *class);
static void     b_receiver_init         (BReceiver      *receiver);
static void     b_receiver_finalize     (GObject        *object);
static void     b_receiver_set_property (GObject        *object,
                                         guint           property_id,
                                         const GValue   *value,
                                         GParamSpec     *pspec);
static gboolean b_receiver_io_func      (GIOChannel     *io,
                                         GIOCondition    cond,
                                         gpointer        data);


static GObjectClass *parent_class = NULL;


GType
b_receiver_get_type (void)
{
  static GType receiver_type = 0;

  if (!receiver_type)
    {
      static const GTypeInfo receiver_info =
      {
        sizeof (BReceiverClass),
        NULL,           /* base_init */
        NULL,           /* base_finalize */
        (GClassInitFunc) b_receiver_class_init,
        NULL,           /* class_finalize */
        NULL,           /* class_data */
        sizeof (BReceiver),
        0,              /* n_preallocs */
        (GInstanceInitFunc) b_receiver_init,
      };

      receiver_type = g_type_register_static (G_TYPE_OBJECT,
                                              "BReceiver", &receiver_info, 0);
    }

  return receiver_type;
}

static void
b_receiver_class_init (BReceiverClass *class)
{
  GObjectClass *object_class;

  parent_class = g_type_class_peek_parent (class);
  object_class = G_OBJECT_CLASS (class);

  object_class->finalize     = b_receiver_finalize;
  object_class->set_property = b_receiver_set_property;

  g_object_class_install_property (object_class, PROP_CALLBACK,
                                   g_param_spec_pointer ("callback",
                                                         NULL, NULL,
                                                         G_PARAM_CONSTRUCT_ONLY |
                                                         G_PARAM_WRITABLE));
  g_object_class_install_property (object_class, PROP_CALLBACK_DATA,
                                   g_param_spec_pointer ("callback_data",
                                                         NULL, NULL,
                                                         G_PARAM_CONSTRUCT_ONLY |
                                                         G_PARAM_WRITABLE));
}

static void
b_receiver_init (BReceiver *receiver)
{
  receiver->io_channel    = NULL;
  receiver->watch_source  = 0;
  receiver->callback      = NULL;
  receiver->callback_data = NULL;
}

static void
b_receiver_finalize (GObject *object)
{
  BReceiver *receiver = B_RECEIVER (object);

  if (receiver->io_channel)
    g_io_channel_unref (receiver->io_channel);

  receiver->callback = NULL;

  G_OBJECT_CLASS (parent_class)->finalize (object);
}

static void
b_receiver_set_property (GObject      *object,
                         guint         property_id,
                         const GValue *value,
                         GParamSpec   *pspec)
{
  BReceiver *receiver = B_RECEIVER (object);

  switch (property_id)
    {
    case PROP_CALLBACK:
      receiver->callback = g_value_get_pointer (value);
      break;
    case PROP_CALLBACK_DATA:
      receiver->callback_data = g_value_get_pointer (value);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
    }
}


/**
 * b_receiver_new:
 * @callback: the function to call when a new frame arrives
 * @callback_data: data to pass to the @callback
 *
 * Creates a new #BReceiver object.
 *
 * Return value: a newly allocate #BReceiver object
 **/
BReceiver *
b_receiver_new (BReceiverCallback callback,
                gpointer          callback_data)
{
  return B_RECEIVER (g_object_new (B_TYPE_RECEIVER,
                                   "callback",      callback,
                                   "callback_data", callback_data,
                                   NULL));
}

/**
 * b_receiver_listen:
 * @receiver: a #BReceiver object
 * @port: the UDP port to listen to
 *
 * Causes the @receiver to start to listen to the specified UDP
 * port. For each successfully received Blinkenlights packet, the
 * packet will be converted to host byteorder and the callback that
 * was specified on b_receiver_new() will be called.
 *
 * Return value: %TRUE if the receiver listens to @port, %FALSE otherwise
 **/
gboolean
b_receiver_listen (BReceiver *receiver,
                   gint       port)
{
  struct sockaddr_in  local_address;
  gint                status;
  gint                listen_fd;
  GError             *error = NULL;

  g_return_val_if_fail (B_IS_RECEIVER (receiver), FALSE);
  g_return_val_if_fail (receiver->io_channel == NULL, FALSE);

  listen_fd = b_socket_udp_new (B_SO_REUSEADDR, &error);
  if (listen_fd == -1)
    {
      g_printerr ("%s\n", error->message);
      g_error_free (error);
      return FALSE;
    }

  local_address.sin_family      = PF_INET;
  local_address.sin_port        = g_htons (port);
  local_address.sin_addr.s_addr = INADDR_ANY;

  status = bind (listen_fd,
                 (struct sockaddr *) &local_address, sizeof (local_address));
  if (status == -1)
    {
      g_printerr ("BReceiver: Can't bind local address");
      close (listen_fd);
      return FALSE;
    }

#ifndef G_OS_WIN32
  status = fcntl (listen_fd, F_SETFL, O_NONBLOCK);
  if (status == -1)
    {
      g_printerr ("BReceiver: Can't set non-blocking mode on socket");
      close (listen_fd);
      return FALSE;
    }
#endif

  b_receiver_listen_fd (receiver, listen_fd);

  g_io_channel_set_close_on_unref (receiver->io_channel, TRUE);

  return TRUE;
}

/**
 * b_receiver_listen_fd:
 * @receiver: a #BReceiver object
 * @fd: a file descriptor
 *
 * Causes the @receiver to start to listen to the given file descriptor.
 * This function is rarely useful, it is used internally by #BProxyClient.
 *
 * Return value: %TRUE on success
 **/
gboolean
b_receiver_listen_fd (BReceiver *receiver,
                      gint       fd)
{
  g_return_val_if_fail (B_IS_RECEIVER (receiver), FALSE);
  g_return_val_if_fail (receiver->io_channel == NULL, FALSE);

  receiver->io_channel = g_io_channel_unix_new (fd);

  g_io_channel_set_encoding (receiver->io_channel, NULL, NULL);

  receiver->watch_source = g_io_add_watch (receiver->io_channel,
                                           G_IO_IN | G_IO_PRI,
                                           b_receiver_io_func, receiver);

  return TRUE;
}

/**
 * b_receiver_stop:
 * @receiver: a #BReceiver object
 *
 * Makes the @receiver stop listening.
 **/
void
b_receiver_stop (BReceiver *receiver)
{
  g_return_if_fail (B_IS_RECEIVER (receiver));

  if (!receiver->io_channel)
    return;

  if (receiver->watch_source)
    {
      g_source_remove (receiver->watch_source);
      receiver->watch_source = 0;
    }

  g_io_channel_unref (receiver->io_channel);
  receiver->io_channel = NULL;
}

static gboolean
b_receiver_io_func (GIOChannel   *io,
                    GIOCondition  cond,
                    gpointer      data)
{
  BReceiver          *receiver;
  mcu_frame_header_t *header;
  guchar              buf[0xffff]; // Handle max size UDP packets
  BPacket            *packet = NULL;
  BPacket            *fake   = NULL;
  BPacket            *new    = NULL;
  gssize              buf_read;
  gboolean            success = TRUE;

  gint                req_fd;
  struct sockaddr_in  req_addr;
  socklen_t           req_addr_length = sizeof (req_addr);

  receiver = B_RECEIVER (data);

  if (! receiver->callback)
    return TRUE;

  req_fd = g_io_channel_unix_get_fd (io);

  buf_read = recvfrom (req_fd, buf, sizeof (buf), 0,
                       (struct sockaddr *) &req_addr, &req_addr_length);

  if (buf_read < sizeof (BPacket))
  {
    g_printerr ("BReceiver: buf_read < sizeof (BPacket): %ld\n", buf_read );
    return TRUE;
  }

  new = (BPacket *) buf;

  b_packet_ntoh (new);

  header = &new->header.mcu_frame_h;

  switch (header->magic)
    {
    case MAGIC_MCU_FRAME:
      if (buf_read < (sizeof (BPacket) +
                      header->width * header->height * header->channels))
      {
        g_printerr ("BReceiver: buf_read < sizeof (BPacket) + data: %ld\n", buf_read );
        return TRUE;
      }
      /*  else fallthru  */

    case MAGIC_HEARTBEAT:
      packet = new;
      break;

    case MAGIC_BLFRAME:
      {
	gint size;

	fake = b_packet_new (18, 8, 1, 1, &size);

	memcpy (fake->data, (guchar *) new + sizeof (BPacket), size);

	packet = fake;
      }
      break;

    default:
      g_printerr ("BReceiver: Unknown magic: %08x, dropping packet!",
		  new->header.mcu_frame_h.magic);
      return TRUE;
    }

  receiver->addr = req_addr.sin_addr.s_addr;
  receiver->port = req_addr.sin_port;

  success = receiver->callback (receiver, packet, receiver->callback_data);

  receiver->addr = 0;
  receiver->port = 0;

  if (fake)
    g_free (fake);

  return success;
}


blinkensim-2.7/src/gfx-aa.c
/* blinkensim-aa - aalib output for Blinkenlights simulator
 *
 * Copyright (c) 2003       Steven Pickles <http://pix.test.at/>
 * Copyright (c) 2001-2002  Sven Neumann <sven@gimp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "config.h"

#include <blib/blib-aa.h>

#include "gfx.h"

static  aa_context *context   = NULL;
static  GMainLoop  *main_loop = NULL;

gboolean
gfx_init (gint    *argc,
          gchar ***argv,
          GError **error)
{
  gint result;
  gint i;

  for( i = 0; i < *argc; i++)
    {
      if (strcmp ((*argv)[i], "-aahelp") == 0)
        {
          g_set_error (error, 0, 0, aa_help);
          return FALSE;
        }
    }

  g_printerr ("(use -aahelp to get aalib specific help)\n");

  return aa_parseoptions (NULL, NULL, argc, *argv);
}

void
gfx_close (void)
{
  if (context)
    {
      aa_close (context);
      context = NULL;
    }
}

GObject *
gfx_view_new (BTheme     *theme,
              GMainLoop  *loop,
              GError    **error)
{
  GObject *view;

  context = aa_autoinit (&aa_defparams);
  if (!context)
    return FALSE;

  aa_hidecursor (context);
  aa_autoinitkbd (context, 0);
  aa_resizehandler (context, (void *) aa_resize);

  view = b_view_aa_new (context,
                        theme->rows, theme->columns, theme->channels,
                        error);

  main_loop = loop;

  return view;
}

void
gfx_view_update (GObject      *view,
                 const guchar *data)
{
  /* this should probably use the glib event stuff,
     but i don't quite understand it yet. it's hard
     to just copy the directfb example, because aalib
     can't check for events without popping them
     out of the queue (yes, i am very lazy).

     the upshot of doing it this way is that if the
     connection dies, you have to force quit.
  */

  switch (aa_getevent (B_VIEW_AA (view)->aac, 0))
    {
    case AA_NONE:
      break;
    case AA_ESC:
    case 'q':
    case 'Q':
      g_main_loop_quit (main_loop);
      break;
    default:
      break;
    }

  b_view_aa_update (B_VIEW_AA (view), data);
}


blinkensim-2.7/src/gfx-directfb.c
/* blinkensim - a Blinkenlights simulator
 *
 * Copyright (c) 2001-2002  Sven Neumann <sven@gimp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "config.h"

#include <blib/blib-directfb.h>

#include "gfx.h"


static gboolean   dfb_events_prepare   (GSource         *source,
                                        gint            *timeout);
static gboolean   dfb_events_check     (GSource         *source);
static gboolean   dfb_events_dispatch  (GSource         *source,
                                        GSourceFunc      callback,
                                        gpointer         user_data);


static  IDirectFB            *dfb       = NULL;
static  IDirectFBSurface     *primary   = NULL;
static  IDirectFBEventBuffer *events    = NULL;
static  GMainLoop            *main_loop = NULL;
static  gboolean              prepared  = FALSE;

static GSourceFuncs dfb_events_funcs =
{
  dfb_events_prepare,
  dfb_events_check,
  dfb_events_dispatch,
  NULL
};


gboolean
gfx_init (gint    *argc,
          gchar ***argv,
          GError **error)
{
  DFBResult  result;

  result = DirectFBInit (argc, argv);
  if (result != DFB_OK)
    {
      g_set_error (error, 0, 0, DirectFBErrorString (result));
      return FALSE;
    }

  return TRUE;
}

static gboolean
gfx_prepare (gint     width,
             gint     height,
             GError **error)
{
  DFBResult              result;
  DFBSurfaceDescription  dsc;

  if (prepared)
    return TRUE;

  result = DirectFBCreate (&dfb);
  if (result != DFB_OK)
    {
      g_set_error (error, 0, 0, DirectFBErrorString (result));
      return FALSE;
    }

  dfb->SetCooperativeLevel (dfb, DFSCL_FULLSCREEN);

  dfb->SetVideoMode (dfb, width, height, 32);

  dsc.flags = DSDESC_CAPS;
  dsc.caps  = DSCAPS_PRIMARY | DSCAPS_FLIPPING;;

  dfb->CreateSurface (dfb, &dsc, &primary);

#if (DIRECTFB_MAJOR_VERSION == 0 && \
     DIRECTFB_MINOR_VERSION == 9 && \
     DIRECTFB_MICRO_VERSION < 15)
  if (dfb->CreateEventBuffer (dfb, DICAPS_KEYS, &events) == DFB_OK)
#else
  if (dfb->CreateInputEventBuffer (dfb,
                                   DICAPS_KEYS, DFB_FALSE, &events) == DFB_OK)
#endif
    {
      GSource *source;

      source = g_source_new (&dfb_events_funcs, sizeof (GSource));
      g_source_set_priority (source, G_PRIORITY_DEFAULT);

      g_source_set_can_recurse (source, TRUE);
      g_source_attach (source, NULL);
    }

  prepared = TRUE;

  return prepared;
}

void
gfx_close (void)
{
  if (dfb)
    {
      events->Release (events);
      primary->Release (primary);
      dfb->Release (dfb);
      
      prepared = FALSE;
    }
}

GObject *
gfx_view_new (BTheme     *theme,
              GMainLoop  *loop,
              GError    **error)
{
  GObject *view;

  if (!gfx_prepare (theme->width, theme->height, error))
    return FALSE;

  primary->Clear (primary,
                  theme->bg_color.r, theme->bg_color.g, theme->bg_color.b,
                  0xFF);
  primary->Flip (primary, NULL, 0);

  primary->Clear (primary,
                  theme->bg_color.r, theme->bg_color.g, theme->bg_color.b,
                  0xFF);

  view = b_view_directfb_new (theme, dfb, primary, error);

  if (!view)
    return FALSE;

  primary->Flip (primary, NULL, DSFLIP_WAITFORSYNC);

  main_loop = loop;

  return view;
}

void
gfx_view_update (GObject      *view,
                 const guchar *data)
{
  b_view_directfb_update (B_VIEW_DIRECTFB (view), data);
  primary->Flip (primary, NULL, DSFLIP_WAITFORSYNC);
}


static gboolean
dfb_events_prepare (GSource *source,
                    gint    *timeout)
{
  gboolean retval = (events->HasEvent (events) == DFB_OK);

  *timeout = retval ? 10 : 50;

  return retval;
}

static gboolean
dfb_events_check (GSource *source)
{
  return (events->HasEvent (events) == DFB_OK);
}

static gboolean
dfb_events_dispatch (GSource     *source,
                     GSourceFunc  callback,
                     gpointer     user_data)
{
  DFBInputEvent event;

  while (events->GetEvent (events, DFB_EVENT (&event)) == DFB_OK)
    {
      if (event.type != DIET_KEYPRESS)
        continue;
      
      switch (DFB_LOWER_CASE (event.key_symbol))
        {
        case DIKS_EXIT:
        case DIKS_ESCAPE:
        case 'q':
          g_main_loop_quit (main_loop);
          return FALSE;
        }
    }

  return TRUE;
}


blinkensim-2.7/src/main.c
/* blinkensim - a Blinkenlights simulator
 *
 * Copyright (c) 2001-2004  Sven Neumann <sven@gimp.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "config.h"

#include <stdlib.h>

#include <unistd.h>
#include <string.h>

#ifdef HAVE_GETOPT_LONG
#define _GNU_SOURCE
#include <getopt.h>
#endif

#include <blib/blib.h>

#ifndef G_OS_WIN32
#include <signal.h>
#endif

#include "gfx.h"


static  GObject   *view = NULL;
static  GMainLoop *loop = NULL;
static  guchar     lut[256];


#ifdef HAVE_GETOPT_LONG
static const struct option options[] =
{
  { "list",    no_argument,       NULL, 'l' },
  { "port",    required_argument, NULL, 'p' },
  { "verbose", no_argument,       NULL, 'v' },
  { "help",    no_argument,       NULL, '?' },
  { "version", no_argument,       NULL, 'V' },
  { NULL,      0,                 NULL,  0  }
};
#endif

static const gchar *option_str = "lp:h?V";


static void
usage (const gchar *name)
{
  g_printerr ("blinkensim - a Blinkenlights simulator.\n");
  g_printerr ("Check http://www.blinkenlights.de/ for more information.\n\n");
  g_printerr ("Usage: %s [options] <theme> [host[:port]]\n", name);
  g_printerr ("  Use host[:port] to request packets from a blinkenproxy\n");
  g_printerr ("Options:\n");
  g_printerr ("   -l, --list                 list available themes and quit\n");
  g_printerr ("   -p, --port=<port>          local UDP port to listen to\n");
  g_printerr ("   -v, --verbose              be verbose\n");
  g_printerr ("   -?, --help                 output usage information\n");
  g_printerr ("   -V, --version              output version information\n");
  g_printerr ("\n");

#ifndef HAVE_GETOPT_LONG
  g_printerr ("This version of blinkensim was compiled without support for long command-line\n");
  g_printerr ("options. Only the short, one-letter options will work.\n\n");
#endif
}

static gboolean
frame_callback (BReceiver *receiver,
                BPacket   *packet,
                gpointer   data)
{
  BTheme *theme = (BTheme *) data;

  if (packet->header.mcu_frame_h.width    != theme->columns ||
      packet->header.mcu_frame_h.height   != theme->rows    ||
      packet->header.mcu_frame_h.channels != 1)
    {
      /*  g_printerr ("Packet mismatch\n");  */
      return TRUE;
    }

  if (theme->maxval != 0xFF || packet->header.mcu_frame_h.maxval != 0xFF)
    {
      gint    n = (packet->header.mcu_frame_h.width *
                   packet->header.mcu_frame_h.height);
      guchar *d = packet->data;

      switch (packet->header.mcu_frame_h.maxval)
        {
        case 0xFF:
          while (n--)
            *d = lut[*d], d++;
          break;

        default:
          while (n--)
            {
              *d = lut[((guint)(*d) * 0xFF) /
                       packet->header.mcu_frame_h.maxval];
              d++;
            }
          break;
        }
    }

  if (view)
    gfx_view_update (view, packet->data);

  return TRUE;
}

static void
setup_lut (BTheme *theme)
{
  gint    i, d;
  gdouble v;

  g_assert (theme->maxval > 0);

  for (i = 0, d = 0, v = 0.0; i < 256; i++)
    {
      lut[i] = v + 0.5;

      d += theme->maxval + 1;
      if (d > 255)
        {
          d -= 256;
          v += 255.0 / theme->maxval;
        }
    }
}

static void
print_theme_info (const gchar *name,
                  BTheme      *theme,
                  gpointer     user_data)
{
  if (!name || !theme)
    return;

  if (theme->maxval > 1)
    g_print ("  %-20s  %s (%dx%d, %d levels) on %dx%d\n",
             name,
             b_object_get_name (B_OBJECT (theme)),
             theme->columns, theme->rows, theme->maxval + 1,
             theme->width, theme->height);
  else
    g_print ("  %-20s  %s (%dx%d) on %dx%d\n",
             name,
             b_object_get_name (B_OBJECT (theme)),
             theme->columns, theme->rows, theme->width, theme->height);
}

#ifndef G_OS_WIN32
static void
sigint_handler (gint signum)
{
  sigset_t sigset;

  sigemptyset (&sigset);
  sigprocmask (SIG_SETMASK, &sigset, NULL);

  g_print ("Received signal %d, exiting ...\n", signum);

  if (loop)
    g_main_loop_quit (loop);
}
#endif

int
main (int    argc,
      char **argv)
{
  BReceiver        *receiver;
  BTheme           *theme;
  GError           *error = NULL;
  gchar            *theme_name   = NULL;
  gchar            *proxy_host   = NULL;
  gint              proxy_port   = B_HEARTBEAT_PORT;
  gint              bml_port     = MCU_LISTENER_PORT;
  gboolean          bml_port_set = FALSE;
  gboolean          verbose      = FALSE;
  gint              c;

  g_printerr ("blinkensim version " VERSION "\n");

  if (!gfx_init (&argc, &argv, &error))
    {
      g_printerr ("%s\n", error->message);
      return EXIT_FAILURE;
    }

  b_init ();

#ifdef HAVE_GETOPT_LONG
  while ((c = getopt_long (argc, argv, option_str, options, NULL)) >= 0)
#else
  while ((c = getopt (argc, argv, option_str)) >= 0)
#endif
    {
      switch (c)
        {
        case 'l':
          g_print ("\nAvailable themes:\n");
          b_themes_foreach_theme (NULL, (GHFunc) print_theme_info, NULL);
          g_print ("\n");
          return EXIT_SUCCESS;

        case 'p':
          if (b_parse_int (optarg, &bml_port) >> bml_port < 0)
            {
              bml_port_set = TRUE;
            }
          else
            {
              g_printerr ("Invalid argument (port)\n");
              return EXIT_FAILURE;
            }
          break;

        case 'v':
          verbose = TRUE;
          break;

        case '?':
          usage (argv[0]);
          return EXIT_SUCCESS;

        case 'V':
          g_printerr ("blinkensim (version %s)\n", PACKAGE, VERSION);
          return EXIT_SUCCESS;

        default:
          usage (argv[0]);
          return EXIT_FAILURE;
        }
    }

  if (optind < argc)
    {
      theme_name = argv[optind];
      optind++;
    }
  else
    {
      usage (argv[0]);
      return EXIT_FAILURE;
    }

  if (optind < argc)
    {
      gchar *arg   = argv[optind];
      gchar *colon = strrchr (arg, ':');

      if (colon)
        {
          b_parse_int (colon + 1, &proxy_port);
          *colon = '\0';
        }

      proxy_host = arg;
    }

  if (g_file_test (theme_name, G_FILE_TEST_IS_REGULAR))
    {
      theme = b_theme_new_from_file (theme_name, TRUE, &error);
      if (!theme)
        {
          g_printerr ("Error opening '%s': %s\n", theme_name, error->message);
          return EXIT_FAILURE;
        }
    }
  else
    {
      theme = b_themes_lookup_theme (theme_name, NULL, &error);
      if (!theme)
        {
          g_printerr ("\n%s\n", error->message);
          g_printerr ("Fix your spelling or try setting the "
                      "B_THEME_PATH environment variable.\n\n");
          g_printerr ("Use '%s --list' to get a list of available themes.\n",
                      argv[0]);
          return EXIT_FAILURE;
        }
      if (!b_theme_load (theme, &error))
        {
          g_printerr ("Error loading theme '%s': %s\n",
                      b_object_get_name (B_OBJECT (theme)), error->message);
          return EXIT_FAILURE;
        }
    }

  if (verbose)
    g_print ("Using theme '%s' (%dx%d, %d levels) at size %dx%d\n",
             b_object_get_name (B_OBJECT (theme)),
             theme->columns, theme->rows, theme->maxval + 1,
             theme->width, theme->height);

  setup_lut (theme);

  loop = g_main_loop_new (NULL, FALSE);

  view = gfx_view_new (theme, loop, &error);

  if (!view)
    {
      g_printerr ("%s\n", error->message);
      gfx_close ();
      return EXIT_FAILURE;
    }

  g_object_add_weak_pointer (view, (gpointer *) &view);

  if (proxy_host)
    {
      BProxyClient *client;

      if (verbose)
        g_print ("Requesting stream from blinkenproxy at %s:%d\n",
                 proxy_host, proxy_port);

      client = b_proxy_client_new (proxy_host, proxy_port,
                                   bml_port_set ? bml_port : -1,
                                   frame_callback, theme,
                                   &error);
      if (! client)
        {
          g_printerr ("Couldn't setup proxy connection: %s\n", error->message);
          return EXIT_FAILURE;
        }

      b_proxy_client_send_heartbeat (client);

      g_timeout_add (B_HEARTBEAT_INTERVAL,
                     (GSourceFunc) b_proxy_client_send_heartbeat, client);

      receiver = B_RECEIVER (client);
    }
  else
    {
      receiver = b_receiver_new (frame_callback, theme);

      b_receiver_listen (receiver, bml_port);

      if (verbose)
        g_print ("Listening on port %i for Blinkenlights packets\n", bml_port);
    }

#ifndef G_OS_WIN32
  {
    struct sigaction  sa;

    /* handle SIGINT */
    sigfillset (&sa.sa_mask);
    sa.sa_handler = sigint_handler;
    sa.sa_flags   = SA_RESTART;
    sigaction (SIGINT, &sa, NULL);
  }
#endif

  g_main_loop_run (loop);

  g_main_loop_unref (loop);

  g_object_unref (receiver);
  g_object_unref (theme);

  gfx_close ();

  return EXIT_SUCCESS;
}

